lib/src/back_end/solution.dart

```diff
@@ -283,15 +283,19 @@ class Solution implements Comparable<Solution> {
 
   @override
   String toString() {
-    var states = _pieceStates.keys
-        .map((piece) => '$piece:${_pieceStates[piece]}')
-        .join(' ');
+    var states = <String>[];
+    _pieceStates.forEach((piece, state) {
+      if (piece.additionalStates.isEmpty) return;
+      if (piece.pinnedState != null) return;
+
+      states.add('$piece${_pieceStates[piece]}');
+    });
```

Comment:
[optional]
```suggestion
    var states = [
      for (var MapEntry(key: piece, value: state) in _pieceStates.entries)
        if (piece.additionalStates.isNotEmpty && piece.pinnedState == null)
          '$piece$state'
    ];
```

Reply:
Oops, I accidentally committed your second suggestion first which collided with this one. I'll commit this locally and push it. Nice suggestion.